app/javascript/packages/rails-i18n-webpack-plugin/rails-i18n-webpack-plugin.js

```diff
@@ -0,0 +1,270 @@
+const { format } = require('util');
+const path = require('path');
+const YAML = require('yaml');
+const ExtractKeysWebpackPlugin = require('./extract-keys-webpack-plugin.js');
+
+/**
+ * @typedef {(key: string, locale: string) => string|undefined|void} MissingStringCallback
+ */
+
+/**
+ * @typedef RailsI18nWebpackPluginOptions
+ *
+ * @prop {string} template String to format using locale data object, generated using util.format.
+ * @prop {string} configPath Root path for locale configuration data.
+ * @prop {string} defaultLocale Default locale to use if string data is missing for desired locale.
+ * @prop {MissingStringCallback} onMissingString Callback to call when a key is missing, optionally
+ * returning a string to use in its place.
+ */
+
+/**
+ * Returns the value in the object at the given key path.
+ *
+ * @example
+ * ```js
+ * const value = dig({ a: { b: { c: 'foo' } } }, ['a', 'b', 'c']);
+ * // 'foo'
+ * ```
+ *
+ * @param {Record<string, any>} object
+ * @param {string[]} keyPath
+ *
+ * @return {any}
+ */
+function dig(object, keyPath) {
+  let result = object;
+  for (const segment of keyPath) {
+    if (result == null) {
+      return;
+    }
+
+    result = result[segment];
+  }
+
+  return result;
+}
+
+/**
+ * Given an array of key-value tuple pairs, returns the corresponding object form.
+ *
+ * @template V
+ *
+ * @param {[key: string, value: V][]} pairs
+ *
+ * @return {Record<string, V>}
+ */
+const fromPairs = (pairs) =>
+  pairs.reduce((result, [key, value]) => ({ ...result, [key]: value }), {});
+
+/**
+ * Returns unique values from the given array.
+ *
+ * @template V
+ *
+ * @param {V[]} values
+ *
+ * @returns {V[]}
+ */
+const uniq = (values) => [...new Set(values)];
+
+/**
+ * Returns truthy values from the given array.
+ *
+ * @template V
+ *
+ * @param {Array<V|undefined>} values
+ *
+ * @returns {V[]}
+ */
+const compact = (values) => /** @type {V[]} */ (values.filter(Boolean));
+
+/**
+ * Returns the given key as a path of parts.
+ *
+ * @param {string} key
+ *
+ * @return {string[]}
+ */
+const getKeyPath = (key) => key.split('.');
+
+/**
+ * Returns domain for a key string, or split key path.
+ *
+ * @param {string|string[]} keyOrKeyPath
+ *
+ * @return {string} The domain.
+ */
+const getKeyDomain = (keyOrKeyPath) =>
+  (Array.isArray(keyOrKeyPath) ? keyOrKeyPath : getKeyPath(keyOrKeyPath))[0];
+
+/**
+ * Returns unique domains for the given set of keys.
+ *
+ * @param {string[]} keys
+ *
+ * @return {string[]}
+ */
+const getKeyDomains = (keys) => uniq(keys.map(getKeyDomain));
+
+/**
+ * @extends {ExtractKeysWebpackPlugin<RailsI18nWebpackPluginOptions>}
+ */
+class RailsI18nWebpackPlugin extends ExtractKeysWebpackPlugin {
+  /** @type {RailsI18nWebpackPluginOptions} */
+  static DEFAULT_OPTIONS = {
+    template: '(window._locale_data = window._locale_data || []).push(%j);',
+    configPath: path.resolve(process.cwd(), 'config/locales'),
+    defaultLocale: 'en',
+    onMissingString: () => {},
```

Comment:
> just to clarify, our default missing behavior will be a completely empty string in development?

It will use the English string if it's present, and only otherwise would it use an empty string.

Reply:
> > just to clarify, our default missing behavior will be a completely empty string in development?
> 
> It will use the English string if it's present, and only otherwise would it use an empty string.

The logic for this behavior is separate from the default:

https://github.com/18F/identity-idp/pull/5223/files#diff-2e1fff2528036656f9d1dd97586d5b36832264ef1d6068240425f6b28f147ec3R189-R191

I could see an alternative implementation where the default option here is what looks up the English string. But I'd think it'd be most common that someone would want to handle `onMissingString` mostly to log or throw, but still use want to take advantage of the English fallback. Returning anything other than `undefined` would prevent that default behavior (even the empty string).

Reply:
Right sorry, yeah I saw the bit about falling back to default locale, but if even the default locale is missing, we get empty string.

Personally when developing I'd like a placeholder, and to not fall back across languages, but also this is fine for now and we can always evolve this if these behaviors don't make sense when developing.

Reply:
Hm, for some reason, I thought that this was the way it worked with Rails by default, but in testing I see it raises an exception if any translation is missing (with `t('foo')`; `I18n.t('foo')` doesn't error, but shows plain text "translation missing: foo").

The English fallback was a source of some headache to implement, so maybe I'll just remove it for now. The tests throwing on missing key should catch missing keys, otherwise I could implement it to fall back to the key, or to a similar message to what Rails does? i.e. `translation missing: ${key}`

Reply:
Rails can be configured with a fallback locale, and the raise on missing is configurable as well.

- We have our set to raise on missing in test and development: https://github.com/18F/identity-idp/blob/9ce2ca623644759a03282999bc9b0198bc8c2475/config/environments/test.rb#L15
- https://github.com/18F/identity-idp/blob/9ce2ca623644759a03282999bc9b0198bc8c2475/config/environments/development.rb#L13

We also have fallbacks true in production 🤷 
https://github.com/18F/identity-idp/blob/main/config/environments/production.rb#L19

Reply:
> Rails can be configured with a fallback locale, and the raise on missing is configurable as well.
> 
> * We have our set to raise on missing in test and development: https://github.com/18F/identity-idp/blob/9ce2ca623644759a03282999bc9b0198bc8c2475/config/environments/test.rb#L15
> * https://github.com/18F/identity-idp/blob/9ce2ca623644759a03282999bc9b0198bc8c2475/config/environments/development.rb#L13
> 
> We also have fallbacks true in production 🤷
> https://github.com/18F/identity-idp/blob/main/config/environments/production.rb#L19

Ah, that explains my confusion then as far as expectations for how it worked.

In that case, I'd agree with your earlier comment: We can iterate on this (maybe with `fallback` and/or `raiseOnMissingTranslations` options) in the future if we need it. I think the current behavior is reasonable enough?

Reply:
Yup yup! Sorry yes I was trying to say the current behavior is definitely fine to ship with for now, as we live with it we might want to change it up.

> Personally when developing I'd like a placeholder, and to not fall back across languages, **but also this is fine for now and we can always evolve this** if these behaviors don't make sense when developing.

LGTM 🚢 it lets goooooo